src/App.tsx

```diff
@@ -1,24 +1,160 @@
+  const [errorText, setErrorText] = useState<null | string>(null);
+
+  const resetError = () => {
+    setErrorText(null);
+  };
+
+  const setError = (errorTxt: string) => {
+    setErrorText(errorTxt);
+    setTimeout(resetError, 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch((error) => {
+        setError(error.message);
+      });
+  }, []);
+
+  const onFilterChange = useCallback((filter: FilterBy) => {
+    setFilterBy(filter);
+  }, []);
+
+  const onAddTodo = async (title: string) => {
+    if (!/\S/g.test(title)) {
+      setError('Title can\'t be empty');
+
+      return false;
+    }
+
+    const newTodo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({ ...newTodo, id: 0 });
+
+    try {
+      const addedTodo = await addTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, addedTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+
+    return true;
+  };
+
+  const onDeleteTodo = useCallback(async (todoId: number) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await deleteTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const onUpdateTodo = useCallback(async (todoId: number, data: TodoPatch) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await updateTodo(todoId, data);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === todoId) {
+          return {
+            ...todo,
+            ...data,
+          };
+        }
+
+        return todo;
+      }));
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const visibleTodos = filterBy === FilterBy.All
+    ? todos
+    : todos.filter(todo => (
+      filterBy === FilterBy.Active
+        ? !todo.completed
+        : todo.completed
+    ));
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <TodoForm
+          todos={todos}
+          onAddTodo={onAddTodo}
+          onUpdateTodo={onUpdateTodo}
+        />
+
+        {Boolean(todos.length) && (
```

Comment:
It is not necessary to cast it directly to the `boolean` type, since dynamic type casting will work here.

Reply:
@roman-mirzoian I don't quite understand this. Should I leave here only todos.length or !!todos.length?

Reply:
@cephy77 you can leave just `{ todos.length && ( ...`
You can read more about types conversions here:
https://www.freecodecamp.org/news/js-type-coercion-explained-27ba3d9a2839